apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/SizedBufferingConsumer.cs

```diff
@@ -61,43 +63,115 @@ private SizedBufferingConsumer(IConsumer<T> consumer, ISizer<T> sizer, int buffe
         /// <param name="consumer">The consumer to flush to, cannot be null.</param>
         /// <param name="bufferSize">The buffer size in bytes.</param>
         public static SizedBufferingConsumer<T> Create(IConsumer<T> consumer, ISizer<T> sizer, int bufferSize)
+            => new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
         {
-            return new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                _semaphore.Wait();
+                foreach (T item in items)
+                {
+                    _size += _sizer.GetSize(item);
+                    _items.Add(item);
+                    if (_size >= _bufferSize)
+                    {
+                        FlushesWithSemaphoreHeld();
+                    }
+                }
+            }
+            finally
+            {
+                _semaphore.Release();
+            }
         }
 
         /// <inheritdoc />
-        public void Receive(IEnumerable<T> items)
+        public async Task ReceiveAsync(
+            IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            lock (_mutex)
+            try
             {
+                await _semaphore.WaitAsync(cancellationToken);
```

Comment:
There's a tiny, tiny risk - which is why `Monitor` changed in .NET 4 and the C# compiler does something slightly different with locks (using a `ref` parameter) but `SemphoreSlim` doesn't support that. I think we should ignore that possibility.